apps/fipsinstall.c

```diff
@@ -89,21 +90,53 @@ static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
     return ret;
 }
 
+static int hexstr2bio(BIO *out, const unsigned char *data, size_t len)
+{
+    size_t i;
+
+    if (len == 0)
+        return 0;
+
+    for (i = 0; i < len; i++) {
+        BIO_printf(out, "%02X", data[i]);
+    }
+    return 1;
+}
+
 static int print_mac(BIO *bio, const char *label, const unsigned char *mac,
                      size_t len)
 {
     if (!BIO_printf(bio, "%s =", label))
         return 0;
-    if (!BIO_hex_string(bio, 1, (int)len, mac, (int)len))
+    if (!hexstr2bio(bio, mac, len))
         return 0;
     return BIO_printf(bio, "\n");
 }
 
-static int load_fips_prov_and_run_self_test(char *fname)
+static int load_fips_prov_and_run_self_test(char *fname,
+                                            const unsigned char *mac,
+                                            size_t mac_len)
 {
+    char buffer[1024], *buf;
+    size_t sz;
     int ret = 0;
     OSSL_PROVIDER *prov = NULL;
     char *prov_name, *dot;
+    static OSSL_PARAM params[2];
+    OSSL_PARAM param_end = OSSL_PARAM_END;
+    BIO *mem_bio = NULL;
+
+    mem_bio = BIO_new_mem_buf(buffer, sizeof(buffer));
+    if (mem_bio == NULL)
+        return 0;
+
+    if (!hexstr2bio(mem_bio, mac, mac_len))
+        goto end;
+
+    params[0] = OSSL_PARAM_construct_utf8_ptr(OSSL_PROV_PARAM_MODULE_MAC,
+                                              &buf, &sz);
+    params[1] = param_end;
+    OSSL_PARAM_set_utf8_ptr(&params[0], buffer);
```

Comment:
That's a pretty convoluted way of saying `buf = buffer`... But now I notice a flaw in `OSSL_PARAM_construct_utf8_ptr`, there's no way to set the size of the current contents...

Reply:
#8703 fixes the constructor so that you can simplify the construction like this:

``` C
    buf = buffer;
    params[0] = OSSL_PARAM_construct_utf8_ptr(OSSL_PROV_PARAM_MODULE_MAC,
                                              &buf, mac_len * 2 + 1, &sz);
    params[1] = param_end;
```

Reply:
Code has been replaced,,